server/curl-adaptor.h

```diff
@@ -0,0 +1,38 @@
+// Compiler for PHP (aka KPHP)
+// Copyright (c) 2023 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+
+#pragma once
+
+#include <memory>
+
+#include "common/smart_ptrs/singleton.h"
+#include "common/mixin/not_copyable.h"
+#include "runtime/signal_safe_hashtable.h"
+
+class CurlRequest;
+class CurlResponse;
+
+struct CurlRequestInfo {
+  int resumable_id{0};
+  std::unique_ptr<CurlResponse> response;
+};
+
+class CurlAdaptor : vk::not_copyable {
+public:
+  int launch_request_resumable(std::unique_ptr<CurlRequest> &&request) noexcept;
+  void process_request_net_query(std::unique_ptr<CurlRequest> &&request) noexcept;
+  void finish_request_resumable(std::unique_ptr<CurlResponse> &&response) noexcept;
+  void process_response_event(std::unique_ptr<CurlResponse> &&response) noexcept;
+
+  std::unique_ptr<CurlResponse> withdraw_response(int request_id) noexcept;
+  void finish_request(int request_id) noexcept;
+  void reset() noexcept;
+
+private:
+  class CurlRequestResumable;
+  int finish_request_processing(int request_id, std::unique_ptr<CurlResponse> &&response) noexcept;
+
+  SignalSafeHashtable<int, std::unique_ptr<CurlRequest>> processing_requests_holder;
```

Comment:
why do we need separate request holders container? It seems that request holder is resumable `curl_exec_concurrently` itself

Reply:
`processing_requests_holder` needed to maintain `CurlRequest` instances. Those instances must be alive till end of curl requests processing they are responsible for. So that callbacks may do safely:
```
auto *curl_request = static_cast<CurlRequest *>(userp);
```

Reply:
As far as I can see, curl request processing starts at the beginning of `curl_exec_concurrently` and ends at the end of it after `finish_request` here
https://github.com/VKCOM/kphp/pull/830/files#diff-86197d94d86236d2c5587aecc9b48ee592f4c12c1ced26be354ee376de2f9d82R33

And none of callbacks related to this curl request will be called after that point. Therefore we can say that lifetime of `curl_exec_concurrently` resumable always includes lifetime of the corresponding curl request processing. Do I miss something?

Reply:
You are right, I've got rid of this map.